Entity point identification and end-to-end functionality

ahemphill opened the pull request.

charbull commented on tools/scoring/score/dimensions/dimension.py:
```diff
@@ -176,8 +180,123 @@ def is_entity_virtual(*, entity: EntityInstance) -> bool:
+      proposed_points_virtual: List[Tuple[Set[str], EntityType]],
+      sort_candidates_by_key: str) -> Dict[float, List[Dict]]:
+    """
+    Finds the closest correlating virtual entities between two files.
```
can you please add details how it is doing that?
tracking through the cloud device id .....

charbull commented on tools/scoring/score/dimensions/dimension.py:
```diff
@@ -176,8 +180,123 @@ def is_entity_virtual(*, entity: EntityInstance) -> bool:
 
+  @staticmethod
+  def match_virtual_entities(
+      *, solution_points_virtual: List[Tuple[Set[str], EntityType]],
```
what is the key here? cloud_device_id?

charbull replied:
do you want to create a structure?
`List[fooStrcture]`?

ahemphill replied:
> what is the key here? cloud_device_id?

Are you asking about `str`? They are raw field names.

> do you want to create a structure?
> `List[fooStrcture]`?

Possibly a good idea — will consider implementation.

charbull commented on tools/scoring/score/dimensions/dimension.py:
```diff
@@ -176,8 +180,123 @@ def is_entity_virtual(*, entity: EntityInstance) -> bool:
+  @staticmethod
+  def match_virtual_entities(
+      *, solution_points_virtual: List[Tuple[Set[str], EntityType]],
+      proposed_points_virtual: List[Tuple[Set[str], EntityType]],
```
same for here

charbull commented on tools/scoring/score/dimensions/dimension.py:
```diff
@@ -176,8 +180,123 @@ def is_entity_virtual(*, entity: EntityInstance) -> bool:
+  def match_virtual_entities(
+      *, solution_points_virtual: List[Tuple[Set[str], EntityType]],
+      proposed_points_virtual: List[Tuple[Set[str], EntityType]],
+      sort_candidates_by_key: str) -> Dict[float, List[Dict]]:
```
it this a score?
what is inside the List[Dict] ?

ahemphill replied:
`float` is a subscore, i.e. how closely the matched entities correlated. The dicts are parameters by which those floats were calculated. (It's described in the docstring but I'll work on annotation.)

charbull commented on tools/scoring/score/dimensions/dimension.py:
```diff
@@ -176,8 +180,123 @@ def is_entity_virtual(*, entity: EntityInstance) -> bool:
+      provided entities correlated and whose values are lists of dictionaries
+      containing the parameters by which those floats were calculated.
+    """
+    matches_virtual = {None: []}
```
please break this down in helper functions

charbull commented on tools/scoring/score/dimensions/dimension.py:
```diff
@@ -176,8 +180,123 @@ def is_entity_virtual(*, entity: EntityInstance) -> bool:
+        if there are multiple matches with the same subscore.
+
+    Returns:
+      Dictionary whose keys are floats representing the extent to which the
```
I am not sure I understand this part.
create a structure?

ahemphill replied:
Good idea.

charbull commented on tools/scoring/score/dimensions/entity_point_identification.py:
```diff
@@ -0,0 +1,169 @@
+  Quantifies whether the proposed file
+  included the correct points in each entity.
+  """
+  def __init__(self, *, deserialized_files: DeserializedFilesDict):
```
please break down into helper functions

charbull commented on tools/scoring/score/dimensions/raw_field_selection.py:
```diff
@@ -29,14 +29,36 @@ class RawFieldSelection(Dimension):
-        map(lambda item: item[1].raw_field_name, translations[PROPOSED]))
+    # Combine translations for all devices within the dictionary
+    solution_condensed = [
+        matched_translations[SOLUTION]
```
this could have a helper function foo(stuff) that we call twice:
foo(SOLUTION)
foo(PROPOSED)

it is the same code repeated, only the input is changing.
what do you think?

ahemphill replied:
I don't have a problem with that other than the fact that it adds an extra assignment for a pretty simply expression — but it does add some degree of protection against dumb mistakes. Will do.

charbull commented on tools/scoring/score/dimensions/standard_field_naming.py:
```diff
@@ -31,17 +31,34 @@ class StandardFieldNaming(Dimension):
+        matched_translations[PROPOSED]
+        for matched_translations in translations.values()
+        if matched_translations[PROPOSED]
+    ]
```
same here, create a helper function, it will reduce the code and make it more readable :)

charbull commented on tools/scoring/score/dimensions/state_mapping.py:
```diff
@@ -28,9 +28,26 @@ class StateMapping(Dimension):
   def __init__(self, *, translations: TranslationsDict):
     super().__init__(translations=translations)
 
+    # Combine translations for all devices within the dictionary
```
helper function?

charbull commented on tools/scoring/score/dimensions/unit_mapping.py:
```diff
@@ -28,16 +28,33 @@ class UnitMapping(Dimension):
+        if matched_translations[PROPOSED]
+    ]
+    # Account for empty list
+    proposed_translations = proposed_condensed and proposed_condensed[0]
```
same, it will reduce your code

charbull commented on tools/scoring/score/parse_config.py:
```diff
@@ -220,3 +236,49 @@ def aggregate_results(
+            standard_field_naming.StandardFieldNaming,
+            state_mapping.StateMapping, unit_mapping.UnitMapping
+        ],
+        COMPLEX: [
```
what is the difference between simple and complex? in the sense, if I want to add a new dimension. How do I know where to add it?

ahemphill replied:
It's noted in `constants.py` and in `dimension.py`. Should I explain here too?

charbull reviewed (changes requested):
can you please refactor into helper functions? it will increase readability of the code.

(didnt check the unit tests yet)

Thank you

charbull commented on tools/scoring/score/dimensions/entity_identification.py:
```diff
@@ -30,38 +30,39 @@ class EntityIdentification(Dimension):
     # reporting entities with canonical types
     solution_reporting = [
         entity.cloud_device_id for entity in filter(
             self.is_entity_canonical,
```
this is the same code, you can use here a helper method, listCloudDeviceId might not be the most appropriate.

```
_listCloudDeviceId(something):
[
        entity.cloud_device_id for entity in filter(
            self.is_entity_canonical,
            filter(self.is_entity_reporting, something.values()))
    ]

```

then you can use it like this:

```
solution_reporting = _listCloudDeviceId(solution_file)
proposed_reporting = _listCloudDeviceId(proposed_file)
```

charbull commented on tools/scoring/score/dimensions/entity_identification.py:
```diff
@@ -30,38 +30,39 @@ class EntityIdentification(Dimension):
     # Lists of `cloud_device_id`s representing
     # reporting entities with canonical types that
     # are linked to by virtual entities
     solution_virtual = [
```
same for this part, use a helper function and use it as above

charbull commented on tools/scoring/score/dimensions/entity_point_identification.py:
```diff
@@ -0,0 +1,148 @@
+        filter(self.is_entity_canonical,
+               filter(self.is_entity_virtual, file.values())))
+
+    solution_entities_virtual = entities_virtual(solution_file)
```
can you please use an action verb on `entities_virtual` 
is get_? `retrieve_virtual_entities`?

move the lambda into a helper function:
```
_get_virtual_entities(file):
return lambda file: set(
        filter(self.is_entity_canonical,
               filter(self.is_entity_virtual, file.values())))

```

charbull commented on tools/scoring/score/dimensions/entity_point_identification.py:
```diff
@@ -0,0 +1,148 @@
+    solution_entities_virtual = entities_virtual(solution_file)
+    proposed_entities_virtual = entities_virtual(proposed_file)
+
+    # For each canonically typed virtual entity
```
put it in a helper function too please

charbull commented on tools/scoring/score/dimensions/entity_point_identification.py:
```diff
@@ -0,0 +1,148 @@
+    solution_entities_reporting = entities_reporting(solution_file)
+    proposed_entities_reporting = entities_reporting(proposed_file)
+
+    # Aggregate IDs for entities which comprise the composites
```
all these lambda should be in helper functions

charbull commented on tools/scoring/score/dimensions/entity_point_identification.py:
```diff
@@ -0,0 +1,148 @@
+  Quantifies whether the proposed file
+  included the correct points in each entity.
+  """
+  def __init__(self, *, deserialized_files: DeserializedFilesDict):
```
in general the `init` must not do operations. 
can you please create another function to do the operations?

ahemphill replied:
I'll have to change all dimensions — will do.

charbull commented on tools/scoring/score/dimensions/state_mapping.py:
```diff
@@ -28,22 +28,27 @@ class StateMapping(Dimension):
+    solution_translations = solution_condensed and solution_condensed[0]
+    proposed_translations = proposed_condensed and proposed_condensed[0]
+
+    mappings = lambda translations: set([(field[0], kv) for field in (
```
same here please

charbull reviewed (changes requested):
please extract the lambda functions in the code into helper functions to increase readability.
I couldnt follow the code :/

ahemphill commented on tools/scoring/score/constants.py:
```diff
@@ -17,16 +17,20 @@
 
 
 class FileTypes(str, Enum):
```
It's inheriting from the built-in.

ahemphill commented on tools/scoring/score/dimensions/dimension.py:
```diff
@@ -176,8 +251,107 @@ def is_entity_virtual(*, entity: EntityInstance) -> bool:
+      candidates = defaultdict(list)
+      # Running tally of which entity most closely corellates;
+      # even zero overlap will best the default at -1.0
+      best: float = -1.1
```
It's only used here.

ahemphill commented on tools/scoring/score/dimensions/dimension.py:
```diff
@@ -176,8 +251,107 @@ def is_entity_virtual(*, entity: EntityInstance) -> bool:
+          best = subscore.tally
+
+        subscore_reference = _VirtualEntityMatch(
+            correct=subscore.correct,
```
> now you can simplify this by providing only the following parameters:
> 
> ```
> _VirtualEntityMatch(subscore, types_score, solution_parameters, proposed_parameters)
> ```

Technically all I need is `correct_ceiling` and `types_score` because they are used to break ties when sorting — this is almost entirely for debugging.

> you can use a cls method ?
> 
> ```
> @classmethod
> def FromScores(subscore, types_score, solution_parameters, proposed_parameters)
> ....
> ```

I don't think I follow.

ahemphill commented on tools/scoring/score/dimensions/dimension.py:
```diff
@@ -176,8 +251,107 @@ def is_entity_virtual(*, entity: EntityInstance) -> bool:
+            types_correct_ceiling=len(
+                set(solution_entity_type.parent_names.keys())),
+            types_incorrect=0,
+            types_score=None)
```
I will pare down the attributes once the other dimension which uses this is implemented.

charbull replied:
ack. can you please add a todo?

ahemphill replied:
ae86f2cc1572e38bd5d23d2fbb4946be00a5eb53

ahemphill replied:
ae86f2cc1572e38bd5d23d2fbb4946be00a5eb53

ahemphill commented:
@charbull Comments addressed

charbull commented on tools/scoring/score/dimensions/dimension.py:
```diff
@@ -13,8 +13,81 @@
+
+
+class _VirtualEntityMatch(NamedTuple):
+  """ Reference for metrics by which subscores were
```
nit, can you please remove those empty spaces and end with a `.`
use thiis pylint style guide? https://google.github.io/styleguide/pylintrc

ahemphill replied:
I'm using this: https://github.com/google/digitalbuildings/blob/master/styles/pylintrc Is it not correct?

ahemphill replied:
https://github.com/google/digitalbuildings/pull/499/commits/68db27e44f32ce704c613c4ce70043dcf9b1ddb9

charbull commented on tools/scoring/score/dimensions/entity_connection_identification.py:
```diff
@@ -27,41 +27,35 @@ class EntityConnectionIdentification(Dimension):
                          for connection in entity.connections)
-                        for cloud_device_id, entity in proposed.items()
+                        for cloud_device_id, entity in file.items()
                         if entity.connections is not None) for tup in tup
```
there are 2x `for tup in tup`  is this legit?
can you please add docstring

ahemphill replied:
Yes, it is correct. (Nested list comprehensions are not the most intuitive.) 40f95b7179a6cd875f731b40685e6c9b24bfeb48

ahemphill replied:
Yes, it is correct. (Nested list comprehensions are not the most intuitive.) 40f95b7179a6cd875f731b40685e6c9b24bfeb48

charbull commented on tools/scoring/score/dimensions/entity_connection_identification.py:
```diff
@@ -27,41 +27,35 @@ class EntityConnectionIdentification(Dimension):
 
+  # TODO: Figure out how to elegantly implement "facilities"
+  # and "equipment" categories given current object model
+  def evaluate(self):
```
can you please add docstring

ahemphill replied:
85c589b8964d292819a342cd0c3dd7006e3705d9

ahemphill replied:
85c589b8964d292819a342cd0c3dd7006e3705d9

charbull commented on tools/scoring/score/dimensions/entity_identification.py:
```diff
@@ -27,51 +29,46 @@ class EntityIdentification(Dimension):
-    self.correct_reporting = sum(
-        (Counter(proposed_reporting) & Counter(solution_reporting)).values())
-    self.correct_ceiling_reporting = len(solution_reporting)
+  def evaluate(self):
```
dostring?

ahemphill replied:
85c589b8964d292819a342cd0c3dd7006e3705d9

ahemphill replied:
85c589b8964d292819a342cd0c3dd7006e3705d9

charbull commented on tools/scoring/score/dimensions/entity_point_identification.py:
```diff
@@ -0,0 +1,159 @@
+  def _fetch_points_virtual(
+      self, file: DeserializedFile,
+      entities_virtual: Set[EntityInstance]) -> PointsVirtualList:
+    # For each canonically typed virtual entity
```
you can use a docstring syntax instead of `#` for comments.

ahemphill replied:
Pylint flags these with `pointless-string-statement`. Should I use the `pylintrc` in the repo root or the one you linked above?

charbull commented on tools/scoring/score/dimensions/entity_point_identification.py:
```diff
@@ -0,0 +1,159 @@
+        for entity in entities_virtual for link in entity.links)
+               for source in sublist)
+
+  def evaluate(self):
```
can you please add docstring here.

ahemphill replied:
85c589b8964d292819a342cd0c3dd7006e3705d9

ahemphill replied:
85c589b8964d292819a342cd0c3dd7006e3705d9

charbull commented on tools/scoring/score/dimensions/entity_point_identification.py:
```diff
@@ -0,0 +1,159 @@
+
+  def evaluate(self):
+
+    ### VIRTUAL ###
```
does it make sense to break this method to two helpers?
virtual and reporting?
as is it is 79 lines of code.

ahemphill replied:
I could go either way.

ahemphill replied:
https://github.com/google/digitalbuildings/pull/499/commits/54ab933ebf45c7f098223b8582eeece47a45a533

charbull commented on tools/scoring/score/parse_config.py:
```diff
@@ -63,19 +65,30 @@ def __init__(self,
-  def append_types(self):
+  @staticmethod
+  def append_types(
+      *, universe: ConfigUniverse,
```
does `*` means any argument?

ahemphill replied:
No, it specifies that all arguments thereafter are keyword-only.

charbull commented on tools/scoring/tests/dimensions/entity_point_identification_test.py:
```diff
@@ -0,0 +1,56 @@
+class EntityPointIdentificationTest(absltest.TestCase):
+  def setUp(self):
+    super().setUp()
+    # TODO: add real data (append cases to existing tests)
```
do you want to start adding real data here?

ahemphill replied:
Yes; I plan to do that just after `entity_type_identification`.

rniemo-g commented on tools/scoring/score/dimensions/standard_field_naming.py:
```diff
@@ -23,33 +22,41 @@
-      s_subfields = set(
-          filter(lambda subfield: not bool(regex.match('[0-9]+', subfield)),
-                 s_field.split('_')))
+    for s_field, s_value in solution_translations:
```
nit: could you rename "s_field" and "s_value" to something like "solution_field" and "solution_value"? Same comment for p_field and p_value below

ahemphill replied:
Good suggestion. https://github.com/google/digitalbuildings/pull/499/commits/b2073f74e5f0bb2c35a7e34a578488bafeb49222

ahemphill commented:
Thanks Ryan — addressed your comment. Are you able to merge this?

On Mon, Mar 14, 2022 at 10:44 AM rniemo-g ***@***.***> wrote:

> ***@***.**** approved this pull request.
> ------------------------------
>
> In tools/scoring/score/dimensions/standard_field_naming.py
> <https://github.com/google/digitalbuildings/pull/499#discussion_r826219686>
> :
>
> >
>      correct_subfields = []
>      correct_ceiling: int = 0
>      incorrect_subfields = []
>
> -    for s_field, s_value in translations[SOLUTION]:
> -      s_subfields = set(
> -          filter(lambda subfield: not bool(regex.match('[0-9]+', subfield)),
> -                 s_field.split('_')))
> +    for s_field, s_value in solution_translations:
>
> nit: could you rename "s_field" and "s_value" to something like
> "solution_field" and "solution_value"? Same comment for p_field and p_value
> below
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/google/digitalbuildings/pull/499#pullrequestreview-909161501>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEUNC6SLNGVOKJY5VRQENTU753GRANCNFSM5OPMVHNA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

rniemo-g commented:
I thinks so. The Github UI shows that Charbel requested some changes that have to be addressed, but when I click into them it looks like you've already addressed them. So I'll ignore all the red text and try merging anyway.
